lib/ansible/module_utils/vmware_pyvmomi.py

```diff
@@ -0,0 +1,688 @@
+#!/usr/bin/python
+
+# -*- coding: utf-8 -*-
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# TODO:
+# Ability to set CPU/Memory reservations
+
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+HAS_PYVMOMI = False
+try:
+    import pyVmomi
+    from pyVmomi import vim
+    from pyVim.connect import SmartConnect, Disconnect
+    HAS_PYVMOMI = True
+except ImportError:
+    pass
+
+import atexit
+import os
+import ssl
+import time
+from pprint import pprint
+
+from ansible.module_utils.urls import fetch_url
+
+
+class PyVmomiHelper(object):
+
+    def __init__(self, module):
+
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='pyvmomi module required')
+
+        self.module = module
+        self.params = module.params
+        self.si = None
+        self.smartconnect()
+        self.datacenter = None
+
+    def smartconnect(self):
+        kwargs = {'host': self.params['vcenter_hostname'],
+                  'user': self.params['username'],
+                  'pwd': self.params['password'],
+                  'port': self.params['vcenter_port']}
+
+        if hasattr(ssl, 'SSLContext'):
+            context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+            context.verify_mode = ssl.CERT_NONE
+            kwargs['sslContext'] = context
+
+        # CONNECT TO THE SERVER
+        try:
+            self.si = SmartConnect(**kwargs)
+        except Exception:
+            err = get_exception()
+            self.module.fail_json(msg="Cannot connect to %s: %s" %
+                             (kwargs['host'], err))
+        atexit.register(Disconnect, self.si)
+        self.content = self.si.RetrieveContent()
+
+    def _build_folder_tree(self, folder, tree={}, treepath=None):
```

Comment:
Is 'tree' given a default mutable arg here intentionally? If so, that would be worth a comment explaining why.

If not then...
``` python
def _build_folder_tree(self, folder, tree=None, treepath=None):
    tree = tree or {'virtualmachines': [],
                             'subfolders': {},
                             'name': folder.name}

```
ditto for 'vmap' on _build_folder_map below.